File: internal/infrastructure/bot/telegram.go
package bot

import (
	"fmt"
	"log"
	"strings"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
	"github.com/robzlabz/angkot/internal/core/services"
	"github.com/robzlabz/angkot/internal/infrastructure/database"
	"github.com/spf13/viper"
)

func Start() error {
	sqlDB, err := database.NewSQLiteDB()
	if err != nil {
		log.Printf("[Adapter][Start]Error initializing SQLite database: %v", err)
		log.Fatal(err)
	}

	// Use SQLite as primary storage
	botService := services.NewBotService(sqlDB)

	bot, err := tgbotapi.NewBotAPI(viper.GetString("TELEGRAM_TOKEN"))
	if err != nil {
		log.Printf("[Adapter][Start]Error initializing Telegram bot: %v", err)
		return err
	}

	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates, err := bot.GetUpdatesChan(u)
	if err != nil {
		log.Printf("[Adapter][Start]Error getting updates channel: %v", err)
		return err
	}

	for update := range updates {
		if update.Message == nil {
			continue
		}

		chatID := update.Message.Chat.ID
		messageText := update.Message.Text

		fmt.Printf("Received message: %s from chat ID: %d\n", messageText, chatID)

		switch {
		case messageText == "/ping":
			response := botService.HandlePing()
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/santri":
			response := botService.HandlePassenger(chatID)
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/daftarsantri":
			response, err := botService.GetPassengerList()
			if err != nil {
				response = "Maaf, terjadi kesalahan saat membaca data santri"
			}
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/driver":
			response := botService.HandleDriver(chatID)
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/drivers":
			response, err := botService.GetDriverList()
			if err != nil {
				response = "Maaf, terjadi kesalahan saat membaca data driver"
			}
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/keberangkatan":
			response := "Format pencatatan keberangkatan:\n\n" +
				"keberangkatan\n" +
				"Driver: [nama_driver]\n" +
				"- [nama_santri_1]\n" +
				"- [nama_santri_2]\n" +
				"- [nama_santri_3]\n\n" +
				"Contoh:\n" +
				"keberangkatan\n" +
				"Driver: Pak Ahmad\n" +
				"- Santri Ali\n" +
				"- Santri Umar\n" +
				"- Santri Hasan"
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/kepulangan":
			response := "Format pencatatan kepulangan:\n\n" +
				"kepulangan\n" +
				"Driver: [nama_driver]\n" +
				"- [nama_santri_1]\n" +
				"- [nama_santri_2]\n" +
				"- [nama_santri_3]\n\n" +
				"Contoh:\n" +
				"kepulangan\n" +
				"Driver: Pak Ahmad\n" +
				"- Santri Ali\n" +
				"- Santri Umar\n" +
				"- Santri Hasan"
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/start" || messageText == "/help":
			response := "Selamat datang! Berikut adalah daftar perintah yang tersedia:\n" +
				"/ping - Cek koneksi bot\n" +
				"/driver - Tambah driver baru\n" +
				"/drivers - Lihat daftar driver\n" +
				"/antar - Lihat format pencatatan antar\n" +
				"/jemput - Lihat format pencatatan jemput\n" +
				"/laporan - Lihat laporan harian"
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/laporan":
			response, err := botService.GetTodayReport()
			if err != nil {
				response = "Maaf, terjadi kesalahan saat membuat laporan"
			}
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/antar":
			response := "Format pencatatan antar:\n\n" +
				"antar\n" +
				"Driver: [nama_driver]\n" +
				"- [nama_santri_1]\n" +
				"- [nama_santri_2]\n" +
				"- [nama_santri_3]\n\n" +
				"Contoh:\n" +
				"antar\n" +
				"Driver: Pak Ahmad\n" +
				"- Santri Ali\n" +
				"- Santri Umar\n" +
				"- Santri Hasan"
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		case messageText == "/jemput":
			response := "Format pencatatan jemput:\n\n" +
				"jemput\n" +
				"Driver: [nama_driver]\n" +
				"- [nama_santri_1]\n" +
				"- [nama_santri_2]\n" +
				"- [nama_santri_3]\n\n" +
				"Contoh:\n" +
				"jemput\n" +
				"Driver: Pak Ahmad\n" +
				"- Santri Ali\n" +
				"- Santri Umar\n" +
				"- Santri Hasan"
			msg := tgbotapi.NewMessage(chatID, response)
			if _, err := bot.Send(msg); err != nil {
				log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
			}
		default:
			if strings.HasPrefix(strings.ToLower(messageText), "antar") {
				lines := strings.Split(messageText, "\n")
				response, err := botService.ProcessDeparture(lines[1], lines[2:])
				if err != nil {
					response = err.Error()
				}
				msg := tgbotapi.NewMessage(chatID, response)
				if _, err := bot.Send(msg); err != nil {
					log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
				}
			} else if strings.HasPrefix(strings.ToLower(messageText), "jemput") {
				lines := strings.Split(messageText, "\n")
				response, err := botService.ProcessReturn(lines[1], lines[2:])
				if err != nil {
					response = err.Error()
				}
				msg := tgbotapi.NewMessage(chatID, response)
				if _, err := bot.Send(msg); err != nil {
					log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
				}
			} else if botService.IsWaitingForPassengerName(chatID) {
				err := botService.AddPassenger(messageText)
				var response string
				if err != nil {
					response = "Maaf, terjadi kesalahan saat menyimpan data penumpang"
				} else {
					response = "Penumpang " + messageText + " berhasil ditambahkan"
				}
				botService.ClearWaitingStatus(chatID)
				msg := tgbotapi.NewMessage(chatID, response)
				if _, err := bot.Send(msg); err != nil {
					log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
				}
			} else if botService.IsWaitingForDriverName(chatID) {
				err := botService.AddDriver(messageText)
				var response string
				if err != nil {
					response = "Maaf, terjadi kesalahan saat menyimpan data driver"
				} else {
					response = "Driver " + messageText + " berhasil ditambahkan"
				}
				botService.ClearWaitingStatus(chatID)
				msg := tgbotapi.NewMessage(chatID, response)
				if _, err := bot.Send(msg); err != nil {
					log.Printf("[Adapter][MessageHandler]Error sending message: %v", err)
				}
			}
		}
	}

	return nil
}

File: src/infrastructure/bot/telegram.ts
import { Bot, Context, InputFile } from "grammy";
import type { BotService } from "../../core/ports/bot";

const HELP_MESSAGE = `
Perintah yang tersedia:
/ping - Cek status bot
/passenger - Tambah penumpang baru
/driver - Tambah supir baru
/passengers - Daftar semua penumpang
/drivers - Daftar semua supir
/report - Laporan hari ini
/report_date YYYY-MM-DD - Laporan per tanggal
/backupdb - Backup database
/catat - Catat perjalanan

Format input perjalanan:
antar
Driver: [nama_supir]
- [penumpang1]
- [penumpang2]

jemput
Driver: [nama_supir]
- [penumpang1]
- [penumpang2]
`;

export function setupBot(bot: Bot, service: BotService): void {
    bot.command("ping", (ctx: Context) => ctx.reply(service.handlePing()));

    bot.command("passenger", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = service.handlePassenger(chatId);
        await ctx.reply(response);
    });

    bot.command("driver", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = service.handleDriver(chatId);
        await ctx.reply(response);
    });

    bot.command("passengers", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = await service.getPassengerList(chatId);
        await ctx.reply(response);
    });

    bot.command("drivers", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = await service.getDriverList(chatId);
        await ctx.reply(response);
    });

    bot.command("report", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = await service.getTodayReport(chatId);
        await ctx.reply(response);
    });

    bot.command("backupdb", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        try {
            const { path, filename } = await service.backupDatabase(chatId);

            // Kirim file database sebagai dokumen
            await ctx.reply("Memulai backup database...");

            await ctx.replyWithDocument(
                new InputFile(path, filename),
                {
                    caption: "Database backup"
                }
            );
        } catch (error) {
            console.error("Error during database backup:", error);
            await ctx.reply("Terjadi kesalahan saat melakukan backup database");
        }
    });

    bot.command("catat", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const response: string = service.handleCatat(chatId);
        await ctx.reply(response);
    });

    bot.on("message", async (ctx: Context) => {
        const chatId: number = ctx.chat?.id ?? 0;
        const message: string | undefined = ctx.message?.text;

        console.log({
            timestamp: new Date().toISOString(),
            chatId: chatId,
            message: message,
            username: ctx.from?.username,
            firstName: ctx.from?.first_name
        });

        if (!message) return;

        if (message.toLowerCase().startsWith('antar')) {
            const inputText = message.substring(5).trim();
            const result = await service.parseAndProcessTrip(inputText, chatId, 'antar');
            await ctx.reply(result);
            return;
        }

        if (message.toLowerCase().startsWith('jemput')) {
            const inputText = message.substring(6).trim();
            const result = await service.parseAndProcessTrip(inputText, chatId, 'jemput');
            await ctx.reply(result);
            return;
        }

        if (service.isWaitingForPassengerName(chatId)) {
            await service.addPassenger(message, chatId);
            service.clearWaitingStatus(chatId);
            await ctx.reply(`Penumpang ${message} berhasil ditambahkan`);
            return;
        }

        if (service.isWaitingForDriverName(chatId)) {
            await service.addDriver(message, chatId);
            service.clearWaitingStatus(chatId);
            await ctx.reply(`Supir ${message} berhasil ditambahkan`);
            return;
        }

        if (service.isWaitingForCatatan(chatId)) {
            const result = await service.processCatatanPerjalanan(message, chatId);
            service.clearWaitingStatus(chatId);
            await ctx.reply(result);
            return;
        }

        if (message.toLowerCase() === '/help' || message.toLowerCase() === 'help') {
            await ctx.reply(HELP_MESSAGE);
            return;
        }
    });

    // Tambahkan command ke daftar commands
    bot.api.setMyCommands([
        // ... command lainnya ...
        { command: "backupdb", description: "Backup database" },
    ]);
}